src/core/thread_pool.cc

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions
+// are met:
+//  * Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+//  * Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of NVIDIA CORPORATION nor the names of its
+//    contributors may be used to endorse or promote products derived
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#include <condition_variable>
+#include <deque>
+#include <future>
+#include <mutex>
+#include <thread>
+#include <vector>
+
+#include "src/core/thread_pool.h"
+
+namespace nvidia { namespace inferenceserver {
+
+WorkerThreadPool::~WorkerThreadPool()
+{
+  cv_.notify_all();
+  for (auto& worker_thread : worker_threads_) {
+    if (worker_thread.joinable()) {
+      worker_thread.join();
+    }
+  }
+}
+
+Status
+WorkerThreadPool::AwaitCompletion()
+{
+  for (size_t i = 0; i < worker_threads_.size(); i++) {
+    if (worker_threads_[i].joinable()) {
+      worker_threads_[i].join();
+      RETURN_IF_ERROR(futures_[i].get());
+    }
+  }
+  return Status::Success;
+}
+
+Status
+WorkerThreadPool::GetNextAvailableId(int* worker_id)
+{
+  *worker_id = -1;
+  for (size_t i = 0; i < worker_threads_.size(); i++) {
+    if (!worker_threads_[i].joinable()) {
+      *worker_id = i;
+    }
+  }
+  if (*worker_id == -1) {
+    RETURN_IF_ERROR(AwaitCompletion());
```

Comment:
If this condition is true, you are returning invalid index for `ProcessQueue()` to use.

Reply:
Yes we should return 0 here after waiting for the threads to complete